### ISSUE TITLE
Event's passing through PQ have their RubyString's encoding changed

### ISSUE THREAD

**Issue by original-brownbear** [elastic/logstash#9167]
The following test for a serialization roundtrip fails:

```java
    @Test
    public void toBinaryRoundtripNonAscii() throws Exception {
        Event e = new Event();
        e.setField("foo", "bör");
        final RubyString before = (RubyString) e.getUnconvertedField("foo");
        Event er = Event.deserialize(e.serialize());
        assertEquals(before, er.getUnconvertedField("foo"));
        assertEquals(
            before.getEncoding(), ((RubyString)er.getUnconvertedField("foo")).getEncoding()
        );
    }
```

with 

```java
java.lang.AssertionError: 
Expected :UTF-8
Actual   :ASCII-8BIT
```

This is causing some real-world issues by breaking e.g. Grok when working with non-ASCII strings (see the linked issue).

This should be fixed by ensuring that the encoding stays constant when going through serialization and deserialization.

**Comment by yaauie** [comment]
@original-brownbear good find.

Since we have a reasonable expectation that strings are UTF-8 for internal representation, and the JSON spec specifically calls out that strings must be unicode, we should ensure that strings are serialised to UTF-8 while building the JSON, and that they are read back out as UTF-8.

I'll put together a patch momentarily.

***

It looks like the `RubyStringDeserializer` being used is deserialising the string using `RubyString#newString(Ruby, byte[])`, which uses the `RubyString(Ruby, RubyClass, byte[])` constructor, creating an ASCII-encoded string backed by a `ByteList`:

> ~~~ java
>         @Override
>         public RubyString deserialize(final JsonParser p, final DeserializationContext ctxt)
>             throws IOException {
>             return RubyString.newString(RubyUtil.RUBY, p.getBinaryValue());
>         }
> ~~~
> -- [logstash-core/src/main/java/org/logstash/ObjectMappers.java:121-125@5dfd0c6](https://github.com/elastic/logstash/blob/5dfd0c60a24b7221bd8e1dc0461c4e16be4d40b2/logstash-core/src/main/java/org/logstash/ObjectMappers.java#L121-L125)

> ~~~ java
>     public static RubyString newString(Ruby runtime, byte[] bytes) {
>         return new RubyString(runtime, runtime.getString(), bytes);
>     }
>  ~~~
> -- [org/jruby/RubyString.java:457-459 jruby/jruby@9.1.12.0](https://github.com/jruby/jruby/blob/9.1.12.0/core/src/main/java/org/jruby/RubyString.java#L457-L459)

> ~~~ java
>     public RubyString(Ruby runtime, RubyClass rubyClass, byte[] value) {
>         super(runtime, rubyClass);
>         assert value != null;
>         this.value = new ByteList(value);
>     }
> ~~~
> -- [org/jruby/RubyString.java:362-366  jruby/jruby@9.1.12.0](https://github.com/jruby/jruby/blob/9.1.12.0/core/src/main/java/org/jruby/RubyString.java#L362-L366)

According to the current standards-track JSON spec, we should be defaulting to UTF-8 when reading the JSON:

> ~~~
> 8.1.  Character Encoding
> 
>    JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32.  The default
>    encoding is UTF-8, and JSON texts that are encoded in UTF-8 are
>    interoperable in the sense that they will be read successfully by the
>    maximum number of implementations; there are many implementations
>    that cannot successfully read texts in other encodings (such as
>    UTF-16 and UTF-32).
> 
>    Implementations MUST NOT add a byte order mark to the beginning of a
>    JSON text.  In the interests of interoperability, implementations
>    that parse JSON texts MAY ignore the presence of a byte order mark
>    rather than treating it as an error.
> ~~~
> -- https://tools.ietf.org/html/rfc7159#section-8.1

**Comment by yaauie** [comment]
I have two alternative approaches, each with their pros and cons:

# [Always Decode as UTF-8](https://github.com/yaauie/logstash/tree/persistent-queue-encoding):

 - PRO: serialisation is unchanged
 - CON: if inputs produce anything _other_ than valid UTF-8, there may be loss, and we do not know what the surface are of this is (e.g., in at least one of our inputs, we ensure that unless explicitly configured to do so, [we don't coerce to UTF-8](https://github.com/logstash-plugins/logstash-input-jdbc/blob/5e6656db8a6b11ab226b2cb449e918e0490a234a/spec/inputs/jdbc_spec.rb#L1001-L1017))

# [Encode as UTF-8 unless Binary](https://github.com/yaauie/logstash/tree/persistent-queue-encoding2):

 - PRO: supports binary-encoded strings side-by-side with UTF-8 and other encodings
 - CON: changes serialisation format; although it is fully backward-compatible, versions of Logstash prior to this change will NOT read `Event`s serialised after this change (makes rollbacks super tricky).

**Comment by original-brownbear** [comment]
@yaauie I would suggest simply serializing the encoding alongside the bytes for `RubyString` for now. If we do anything like the above suggestions, we're simply creating behavior differences between the PQ and normal queue which will probably just lead to more unexpected issues down the road like this one.
Why not simply prefix the bytes with the encoding in `org.logstash.ObjectMappers.RubyStringSerializer` (I'm happy to spell out some suggestions on how to implement this efficiently if you want some)? That should work fine right?

**Comment by yaauie** [comment]
> If we do anything like the above suggestions, we're simply creating behavior differences between the PQ and normal queue which will probably just lead to more unexpected issues down the road like this one.

@original-brownbear: that is a good point; after taking a step back, it would make sense that _if_ we were to start enforcing UTF-8 all-the-way-through, we should do it in the `org.logstash.Valuifier` that is used for setting values on an `org.logstash.Event`, and that _may_ be out of scope.

 - From offline discussions with @jordansissel, I understand that there is an implicit expectation that the `RubyString`s in `Event`s produced by input plugins and codecs will always be valid UTF-8; unicode can represent every possible glyph in every supported encoding, so it is a safe intermediate encoding for strings that represent sequences of glyphs (this is why Java `String`s are Unicode all the way through).
 - Unicode, however, does not provide a way to support arbitrary byte sequences; while in Java we simply have `byte[]`, on the Ruby side of the fence, we represent these as a `RubyString` with a `BINARY` encoding (which is just an alias for `ASCII-8BIT`, meaning "any sequence is legal" and "exactly one byte per char").

I feel we have three options here:
 1. actually enforce UTF-8 as the _only_ intermediate `RubyString` encoding (thereby blocking the use of `BINARY`-encoded `RubyString`; a change in the `org.logstash.Valuefier` could enforce this; I believe this to be a "breaking change")
 2. coerce to `UTF-8` wherever we are guaranteed to not be lossy (which is all `RubyString`s except those encoded with `BINARY`/`ASCII-8BIT`; could also be done in the `org.logstash.Valuefier`, but would also require backward-compatible serialisation changes similar to [Encode as UTF-8 unless Binary](https://github.com/elastic/logstash/compare/master...yaauie:persistent-queue-encoding2), without the fallback to convert to UTF-8) to ensure we also survive serialisation round-trip
 3. propagate `RubyString` encoding through serialisation by means of a marker (however: I can't think of a backward-compatible way to do this; since the field is already being stored with unsafe-bytes, there is no way to inject a magic byte-sequence to indicate an encoding in such a way that we could make decisions based on its absence)

**Comment by original-brownbear** [comment]
@yaauie 

> I understand that there is an implicit expectation that the RubyStrings in Events produced by input plugins and codecs will always be valid UTF-8

Not sure we can actually make that assumption for Ruby plugins. I'm actually 90% certain this doesn't hold for JRuby code on Windows for certain locale configurations.

That said, I said I can give some pointers on how to implement this :)
=> I think the easiest way of handling the BwC is to:

1. Use an invalid UTF-8 byte sequence to mark the fact we are writing encoding information
2. Assume UTF-8 in the deserializer (for existing data in the queue)
3. If the `byte[]` we deserialize starts on the non-UTF-8 prefix from 1. => read the serialized encoding from the next few bytes
   * I think the performance hit form this is irrelevant. If implement properly it's just the performance of reading a `long` from a known memory address.
   * I know that the extreme corner case of an encoding that has the prefix we chose in 1. as a valid prefix in it would still be broken but since that's broken right now anyway due to this bug, so it doesn't matter imo.
4. Profit :)

**Comment by wwalker0307** [comment]
Has any progress been made on this?  Any chance there's a hotfix I can download to test/use?  I've got an implementation utilizing the XML filter which does not play nice with ASCII encoding.

**Comment by original-brownbear** [comment]
@yaauie as discussed dealing with this now :)

**Comment by original-brownbear** [comment]
https://github.com/original-brownbear/logstash/commit/5095e7b7f9f38d79ae0f3faf1b41c6e830fd66e9 fixes this just fine using the approach outlined above and is BwC (could be optimized a little more though). Will put up a more optimal PR that uses the approach after the holidays.

**Comment by original-brownbear** [comment]
fixed in #9307 :)
